### PR TITLE
11.0 mig delivery auto refresh

### PR DESCRIPTION
**Pull request by chienandalu**
Backporting from v12 as most changes apply already to the v11 realm

cc @Tecnativa TT22689

### CODE REVIEW

File: delivery_auto_refresh/models/stock_picking.py [hunk 1]
```diff
@@ -34,9 +34,15 @@ def _add_delivery_cost_to_so(self):
...
-            sale_order.date_order)
+        # In v11 there's no _convert method to get proper currency conversion
+        # according to a given date. We replicate such behavior here.
+        if sale_order.company_id.currency_id != sale_order.currency_id:
+            to_currency = sale_order.company_id.currency_id
+            from_currency = sale_order.currency_id
+            total = total * from_currency._get_conversion_rate(
+                from_currency, to_currency,
+                sale_order.company_id, sale_order.date_order)
+            total = to_currency.round(total)
```
Comment (pedrobaeza) [inline delivery_auto_refresh/models/stock_picking.py]: You have to use instead the method `compute`:
```python
total = sale_order.currency_id.with_context(date=sale_order.date_order).compute(total, sale_order.company_id.currency_id)
```

**Review by pedrobaeza — CHANGES_REQUESTED** [review]
See comment

File: delivery_auto_refresh/models/stock_picking.py (hunk 1 above)
Comment (chienandalu) [reply delivery_auto_refresh/models/stock_picking.py]: :+1:

**Comment by pedrobaeza** [comment]
Travis is failing

**Comment by pedrobaeza** [comment]
Sorry, still red

**Comment by chienandalu** [comment]
Ok, green now. I had to revert some of the changes of this commit, since they doesn't seem compatible with v11: https://github.com/OCA/delivery-carrier/commit/c71e659c595cec900a810c4aab3c2c9816a3068c#diff-0e55f8aa64b3c27b1a13c770288e3edcR11-R35

File: delivery_auto_refresh/models/sale_order.py [hunk 2]
```diff
@@ -1,36 +1,33 @@
...
+                self.with_context(force_delivery_set=True).set_delivery_line()
 
     def write(self, vals):
-        """Create or refresh delivery line after saving."""
+        """Refresh delivery price after saving."""
+        vals.update({'delivery_rating_success': True})
         res = super(SaleOrder, self).write(vals)
         for order in self:
-            order._auto_refresh_delivery()
+            force = order.order_line.filtered('is_delivery')
```
Comment (pedrobaeza) [inline delivery_auto_refresh/models/sale_order.py]:
```suggestion
            force = bool(order.order_line.filtered('is_delivery'))
```

File: delivery_auto_refresh/tests/test_delivery_auto_refresh.py [hunk 3]
```diff
@@ -11,6 +11,8 @@ def _execute_onchanges(records, field_name):
             onchange(record)
 
 
+@common.at_install(False)
```
Comment (pedrobaeza) [inline delivery_auto_refresh/tests/test_delivery_auto_refresh.py]: Can we avoid this?

**Review by pedrobaeza — APPROVED** [review]

File: delivery_auto_refresh/tests/test_delivery_auto_refresh.py (hunk 3 above)
Comment (chienandalu) [reply delivery_auto_refresh/tests/test_delivery_auto_refresh.py]: Let's try

Comment (chienandalu) [reply delivery_auto_refresh/tests/test_delivery_auto_refresh.py]: works fine

File: delivery_auto_refresh/README.rst [hunk 4]
```diff
@@ -0,0 +1,111 @@
...
+
+Authors
+~~~~~~~
+
+* Tecnativa
+
+Contributors
+~~~~~~~~~~~~
+
+* Tecnativa <https://www.tecnativa.com>:
```
Comment (ernestotejeda) [inline delivery_auto_refresh/README.rst:91]:
```suggestion
* `Tecnativa <https://www.tecnativa.com>`_:

```
Add also the empty line to be displayed correctly

**Review by rafaelbn — APPROVED** [review]
Functionally tested :+1:

**Comment by rousseldenis** [comment]
@chienandalu This one ready ?

**Comment by chienandalu** [comment]
@rousseldenis we just detected an issue reported from a customer (see last commit), so we'll try to implement it in this same pr and then forward port it to 12.0

File: delivery_auto_refresh/models/sale_order.py [hunk 5]
```diff
@@ -25,9 +25,21 @@ def write(self, vals):
         vals.update({'delivery_rating_success': True})
         res = super(SaleOrder, self).write(vals)
         for order in self:
-            force = bool(order.order_line.filtered('is_delivery'))
+            delivery_line = order.order_line.filtered('is_delivery')
             # Make sure that if you have removed the carrier, the line is gone
+            discount = delivery_line and delivery_line.discount or 0
```
Comment (pedrobaeza) [inline delivery_auto_refresh/models/sale_order.py]:
```suggestion
            discount = delivery_line.discount
```

File: delivery_auto_refresh/models/sale_order.py [hunk 6]
```diff
@@ -25,9 +25,21 @@ def write(self, vals):
...
+                delivery_discount=discount
+                )._auto_refresh_delivery(force=bool(delivery_line))
         return res
+
+    def _create_delivery_line(self, carrier, price_unit):
+        """Allow users to keep discounts to delivery lines. Unit price will
+           be recomputed anyway"""
+        sol = super()._create_delivery_line(carrier, price_unit)
+        discount = self.env.context.get('delivery_discount')
+        if discount:
```
Comment (pedrobaeza) [inline delivery_auto_refresh/models/sale_order.py]:
```suggestion
        if discount and sol:
```

**Comment by pedrobaeza** [comment]
Cherry-picked commit about discount in 6a01ca619eb00825b3ba2a96655e8adb3c3fd1a7 for v12

**Pull request merged**